# SNLP Assignment 2

Name 1: Aleksey Morshnev<br/>
Student id 1: 7042691<br/>
Email 1: almo00008@stud.uni-saarland.de<br/>


Name 2: <br/>
Student id 2: <br/>
Email 2: <br/> 

Name 3: <br/>
Student id 3: <br/>
Email 3: <br/> 

**Instructions:** Read each question carefully. <br/>
Make sure you appropriately comment your code wherever required. Your final submission should contain the completed Notebook. There is no need to submit the data files. <br/>
Upload the zipped folder on CMS. Please follow the naming convention of **Name1_studentID1_Name2_studentID2_Name3_studentID3.zip**. Make sure to click on "Turn-in" (or the equivalent on CMS) after you upload your submission, otherwise the assignment will not be considered as submitted. Only one member of the group should make the submisssion.

---

## Exercise 1 - Probability theory (0.5 + 0.5 = 1 point)

Consider a fair 6-sided die whose sides are numbered from 1 to 6 and each die roll is independent of the other rolls. In an experiment that consists of rolling the die twice, the following events can be defined:

    A: The sum of the two outcomes is at least 10
    B: At least one of the two rolls resulted in 6

a. Compute the probabilities $P(A)$, and $P(B)$.

b. Is event A independent of event B?

You can leave the solution as a fraction.

---

#### Solution 1

## Exercise 2 - Bayes theorem (0.5 + 1 = 1.5 points)

Bayes' theorem is a fundamental concept in probability theory used to update the probability of a hypothesis based on new evidence. It states that the probability of a hypothesis (H) given some evidence (E) is proportional to the probability of the evidence given the hypothesis, multiplied by the prior probability of the hypothesis, divided by the probability of the evidence regardless of the hypothesis. Mathematically,

$$ P(H|E) = \frac{P(E|H)P(H)}{P(E)} $$

Where,
- $P(H∣E)$ is the posterior probability of hypothesis H given evidence E.
- $P(E∣H)$ is the probability of observing evidence E given that the hypothesis H is true.
- $P(H)$ is the prior probability of hypothesis H being true.
- $P(E)$ is the probability of observing evidence E.


Now, Imagine that all of the blogs online were generated by some Large Language Model (LLM). Of these blogs, $80\%$ were generated using GPT-4, $15\%$ of them were generated using LLAMA 3 and $5\%$ were generated using Mistral. About $4\%$ of the text generated by GPT-4 seem to have counterfactual statements. For LLAMA 3, this is about $6\%$ and for Mistral, this is about $9\%$.

a. If a blog post is randomly selected from these online blogs, what is the probability that it was generated by LLAMA 3?

b. If a randomly selected blog post contained a counterfactual statement, find the probability that it was generated using GPT-4.

---

#### Solution 2

## Exercise 3 - Miller's model (0.5 + 1 + 0.5 = 2 points)

Miller's model introduced in the lecture simulates the process of random text generation at the character level. It is a memoryless model which means that the generation at each time step is independent of what has been generated so far. Imagine that there is a monkey typing on a computer keyboard. The keyboard has only 27 keys: a to z and a spacebar. Assume that the probability of each character is distributed according to its occurrence in the English language. We call a sequence of characters separated by white space as a word.

a. What is the probability that the monkey will type the word _'ear'_ on the keyboard? Hint: See the `exercise_3.py` file.

b. Complete the `compute_perplexity()` function in the `exercise_3.py` file to compute the perplexity of the Miller's model for the text: "hello i am a monkey". 

c. What would happen to the perplexity if we were to add a few more whitespace in between the words?

---

#### Solution 3

In [ ]:
# Solution 3.b

# DO NOT MODIFY THIS PART
# Your code should be placed in exercise_3.py
import exercise_3
from importlib import reload

reload(exercise_3)

model = exercise_3.MillersModel()

text = "hello i am a monkey"

print("Perplexity: ", model.compute_perplexity(text))

## Exercise 4 - Sampling in Language models (1 + 2 + 0.5 + 0.5 =4 points)

Neural Networks are quite effective as generative models. Huggingface Transformers library can be used to easily get started with training and using these models. Take a look at the [Huggingface transformers quick tour](https://huggingface.co/docs/transformers/quicktour) to get acquainted with the transformers library. (Recommended: [Chapter 1](https://huggingface.co/learn/nlp-course/chapter1/1) and [Chapter 2](https://huggingface.co/learn/nlp-course/chapter2/1) of the Huggingface NLP Course)

a. Read up on GPT2 [[1]](https://huggingface.co/openai-community/gpt2) [[2]](https://huggingface.co/docs/transformers/model_doc/gpt2). Implement the `forward()` function in the `GPT2Model` class in `exercise_4.py` file to take input text sequence and return the probability distribution of the predicted next token. The general steps would be: tokenize and encode the input text, feed the input to the model to get the logits (Note: You might have to take only the last logit which is for the predicted next token in the sequence.), and apply softmax function to convert it to a probability distribution.

b. We can use different types of sampling strategies while selecting the next token for generating a text. Some common approaches are: Greedy sampling, random sampling, beam search, etc.

   - Greedy sampling
      In greedy sampling, the model selects the word with the highest probability based on the given context. It prioritizes the most likely next word based on the model's predictions, without considering other possibilities. This method often produces coherent and fluent text but can lead to repetitive or predictable outputs, as it doesn't explore alternative word choices.

   - Random sampling
      In random sampling, the next token is randomly sampled from the output probability distribution. This can make the generation non-deterministic and unpredictable but doesn't usually fall into repetitions.

   Implement the `greedy_sample()` and `random_sample()` functions in `GPT2Model` class in `exercise_4.py` file for Greedy sampling and random sampling respectively. For each of them, generate a continuation of the sentence given below.

c. Describe the output with each sampling method. Which one would you generally prefer?

d. Another approach for generating from language models is using Beam search. Describe how the beam search works (2 sentences). Does it solve any problems of the previous sampling methods?

---

In [3]:
# install necessary packages
!conda install -y transformers evaluate datasets

done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 23.7.4
  latest version: 24.5.0

Please update conda by running

    $ conda update -n base -c conda-forge conda

Or to minimize the number of packages updated during conda update use

     conda install conda=24.5.0



## Package Plan ##

  environment location: /home/reni/miniconda3/envs/hws_ws23

  added / updated specs:
    - datasets
    - evaluate
    - transformers


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    arrow-1.3.0                |     pyhd8ed1ab_0          98 KB  conda-forge
    binaryornot-0.4.4          |             py_1         370 KB  conda-forge
    cookiecutter-2.6.0         |     pyhca7485f_0          98 KB  conda-forge
    datasets-2.19.1            |     pyhd8ed1ab_0         350 KB  conda-forge
    evaluate-0.4.1             |     pyhd8ed1ab_0          61 KB 

In [21]:
# Solution 4

# DO NOT MODIFY THIS PART
# Your code should be placed in exercise_4.py
import exercise_4
from importlib import reload

reload(exercise_4)

model = exercise_4.GPT2Model()

In [24]:
# Greedy sampling
n_tokens = 10

text = "Alice was beginning to get very tired of sitting by her sister on the bank "

for _ in range(n_tokens):
    text += model.greedy_sample(text)

print("Greedy sampled output: ", text)

Greedy sampled output:  Alice was beginning to get very tired of sitting by her sister on the bank   and she was starting to get tired of sitting


In [29]:
# Random sampling
n_tokens = 10

text = "Alice was beginning to get very tired of sitting by her sister on the bank "

for _ in range(n_tokens):
    text += model.random_sample(text)

print("Random sampled output: ", text)

Random sampled output:  Alice was beginning to get very tired of sitting by her sister on the bank   while she took herself up on the offer.


## Exercise 5 - Computing Perplexity (0.5 + 1 = 1.5 points)

Huggingface provides the `evaluate` library that includes various metrics used in machine learning along with perplexity that we use in Language modeling. We can use that library to quickly compute the perplexity of a language model. There is also the `datasets` library from Huggingface that allows to easily access datasets for many different tasks.

a. Read the [metric card for perplexity](https://huggingface.co/spaces/evaluate-metric/perplexity) and compute the perplexity for the following text using GPT2.

b. Load the `alice_in_wonderland.txt` file using Hugginface `datasets` library and compute the perplexity for it using GPT2. Check the [documentations](https://huggingface.co/docs/datasets/index) for datasets library.

---

In [33]:
# Solution 5.a
from evaluate import load

texts = ["Eureka! I have found it!"]

# your code here
perplexity = load("perplexity", module_type="metric")
results = perplexity.compute(predictions=texts, model_id='gpt2')

print("Perplexity: ", results['mean_perplexity'])

  0%|          | 0/1 [00:00<?, ?it/s]

Perplexity:  24.384519577026367


In [43]:
# Solution 5.b
from transformers import GPT2Tokenizer
from datasets import load_dataset
from evaluate import load

# Load the tokenizer
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')

# Load the dataset
dataset = load_dataset('text', data_files={'train': 'alice_in_wonderland.txt'}, split='train')

# Function to check token length
def check_token_length(text):
    tokens = tokenizer.encode(text, add_special_tokens=True)
    if len(tokens) < 1:  # Checks if the token length is less than 1
        return False
    return True

# Filter and extract text data ensuring each line results in at least one token
valid_texts = []
for item in dataset:
    text = item['text'].strip()
    if text and check_token_length(text):
        valid_texts.append(text)

perplexity = load("perplexity", module_type="metric")
results = perplexity.compute(predictions=valid_texts, model_id='gpt2')

# map the perplexities to the corresponding text
# perplexities = dict(zip(valid_texts, results['perplexities']))
# print(perplexities)

print("Mean Perplexity: ", results['mean_perplexity'])

  0%|          | 0/2 [00:00<?, ?it/s]

{'Alice was beginning to get very tired of sitting by her sister': 60.3527717590332, 'on the bank, and of having nothing to do:  once or twice she had': 181.5369110107422, 'peeped into the book her sister was reading, but it had no': 97.06128692626953, "pictures or conversations in it, `and what is the use of a book,'": 165.4005126953125, "thought Alice `without pictures or conversation?'": 9847.4453125, 'So she was considering in her own mind (as well as she could,': 70.94112396240234, 'for the hot day made her feel very sleepy and stupid), whether': 401.9328308105469, 'the pleasure of making a daisy-chain would be worth the trouble': 51.625770568847656, 'of getting up and picking the daisies, when suddenly a White': 152.7297821044922, 'Rabbit with pink eyes ran close by her.': 256.022705078125, 'There was nothing so VERY remarkable in that; nor did Alice': 285.7662658691406, 'think it so VERY much out of the way to hear the Rabbit say to': 326.69989013671875, "itself, `Oh dear!  Oh d

## Bonus

## Rejection Sampling (2 points)

Rejection sampling is another approach we can use to generate the next token from the language model. In this approach, we randomly sample a token from the list and accept it if it's probability is greater than some random number and reject it otherwise.

1. Generate a random number, r.
2. Randomly sample a token from the output probability (say token at index i with probability p).
3. Is r < p?

    a. If yes, then accept i as the next token.
    
    b. If no, then go back to step 1.

Complete the `rejection_sample()` function in the `GPT2Model` class in `exercise_4.py`. How does the generated output differ from the greedy and random sampling?

---

Answer:

Rejection Sampling method introduces a balance between randomness and probability weighted choices. Output is a compromise between random sampling and greedy sampling.

In [31]:
import exercise_4
from importlib import reload
reload(exercise_4)

# Random sampling
n_tokens = 10

text = "Alice was beginning to get very tired of sitting by her sister on the bank "
model = exercise_4.GPT2Model()

for _ in range(n_tokens):
    text += model.random_sample(text)

print("Rejection sampled output: ", text)

Rejection sampled output:  Alice was beginning to get very tired of sitting by her sister on the bank   until her sisters arrived and struggled to free her


## Perplexity with different languages (0.5 + 0.5 = 1 point)

a. In exercise 5.a., what would happen if we were to use a different language other than English? How would the perplexity be affected? Test this with another language of your choice.

b. Are there any pre-trained model for the language you selected? Try with that model to see if you can get a better perplexity score. If there are no pre-trained LM for that language, try with a multilingual LM and see if the perplexity score improves.

Answer:

a. GPT-2 was primarily trained on English text, so predicting the next token in a different language results in a much higher perplexity score.

In [46]:
from evaluate import load

# Load the perplexity metric
perplexity = load("perplexity", module_type="metric")

# Uzbek text to compute perplexity
uzbek_text = ["Evrika! Men topdim!"]

# Compute the perplexity using the GPT-2 model
results = perplexity.compute(predictions=uzbek_text, model_id='gpt2')

# Print the perplexity result
print("Perplexity for Uzbek text: ", results['mean_perplexity'])


  0%|          | 0/1 [00:00<?, ?it/s]

Perplexity for Uzbek text:  41356.34375


Answer:

b. There is a pretrained model for Uzbek language called "mGPT-1.3B-uzbek". Using it, the perplexity score decresed from 41356 to 301.

In [50]:
from evaluate import load

# Load the perplexity metric
perplexity = load("perplexity", module_type="metric")

# Uzbek text to compute perplexity
uzbek_text = ["Evrika! Men topdim!"]

# Compute the perplexity using the GPT-2 model
results = perplexity.compute(predictions=uzbek_text, model_id='ai-forever/mGPT-1.3B-uzbek')

# Print the perplexity result
print("Perplexity for Uzbek text: ", results['mean_perplexity'])


pytorch_model.bin:   0%|          | 0.00/5.77G [00:00<?, ?B/s]

Error while downloading from https://cdn-lfs.huggingface.co/repos/7b/41/7b41f25cbb46d968cf00011078b58c0722841346a3fc3fc65dc24c3a472be3f5/5b742844040495fcbb9db2affe548c5f28b4a22ac27b9a2ea047c4df19fd17e7?response-content-disposition=attachment%3B+filename*%3DUTF-8%27%27pytorch_model.bin%3B+filename%3D%22pytorch_model.bin%22%3B&response-content-type=application%2Foctet-stream&Expires=1716319759&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTcxNjMxOTc1OX19LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5odWdnaW5nZmFjZS5jby9yZXBvcy83Yi80MS83YjQxZjI1Y2JiNDZkOTY4Y2YwMDAxMTA3OGI1OGMwNzIyODQxMzQ2YTNmYzNmYzY1ZGMyNGMzYTQ3MmJlM2Y1LzViNzQyODQ0MDQwNDk1ZmNiYjlkYjJhZmZlNTQ4YzVmMjhiNGEyMmFjMjdiOWEyZWEwNDdjNGRmMTlmZDE3ZTc%7EcmVzcG9uc2UtY29udGVudC1kaXNwb3NpdGlvbj0qJnJlc3BvbnNlLWNvbnRlbnQtdHlwZT0qIn1dfQ__&Signature=RLlvVNgjZD5bnFosmuhvWZyOrE-gI-j8WGCX%7E5D4wu%7EfDAuuQfMbESYSwP0Z8aK%7Eru6-XIAVtTjCXZ4xSNSQc7Y4q6ksVBnyQBSPdpC9HKiCf5ZVbegriqfUcmXBnpO8kFR8NDJDbtxZ1H-uJAsqWHJVsgz

pytorch_model.bin:   1%|1         | 62.9M/5.77G [00:00<?, ?B/s]

Error while downloading from https://cdn-lfs.huggingface.co/repos/7b/41/7b41f25cbb46d968cf00011078b58c0722841346a3fc3fc65dc24c3a472be3f5/5b742844040495fcbb9db2affe548c5f28b4a22ac27b9a2ea047c4df19fd17e7?response-content-disposition=attachment%3B+filename*%3DUTF-8%27%27pytorch_model.bin%3B+filename%3D%22pytorch_model.bin%22%3B&response-content-type=application%2Foctet-stream&Expires=1716319759&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTcxNjMxOTc1OX19LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5odWdnaW5nZmFjZS5jby9yZXBvcy83Yi80MS83YjQxZjI1Y2JiNDZkOTY4Y2YwMDAxMTA3OGI1OGMwNzIyODQxMzQ2YTNmYzNmYzY1ZGMyNGMzYTQ3MmJlM2Y1LzViNzQyODQ0MDQwNDk1ZmNiYjlkYjJhZmZlNTQ4YzVmMjhiNGEyMmFjMjdiOWEyZWEwNDdjNGRmMTlmZDE3ZTc%7EcmVzcG9uc2UtY29udGVudC1kaXNwb3NpdGlvbj0qJnJlc3BvbnNlLWNvbnRlbnQtdHlwZT0qIn1dfQ__&Signature=RLlvVNgjZD5bnFosmuhvWZyOrE-gI-j8WGCX%7E5D4wu%7EfDAuuQfMbESYSwP0Z8aK%7Eru6-XIAVtTjCXZ4xSNSQc7Y4q6ksVBnyQBSPdpC9HKiCf5ZVbegriqfUcmXBnpO8kFR8NDJDbtxZ1H-uJAsqWHJVsgz

pytorch_model.bin:  37%|###7      | 2.16G/5.77G [00:00<?, ?B/s]

Error while downloading from https://cdn-lfs.huggingface.co/repos/7b/41/7b41f25cbb46d968cf00011078b58c0722841346a3fc3fc65dc24c3a472be3f5/5b742844040495fcbb9db2affe548c5f28b4a22ac27b9a2ea047c4df19fd17e7?response-content-disposition=attachment%3B+filename*%3DUTF-8%27%27pytorch_model.bin%3B+filename%3D%22pytorch_model.bin%22%3B&response-content-type=application%2Foctet-stream&Expires=1716319759&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTcxNjMxOTc1OX19LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5odWdnaW5nZmFjZS5jby9yZXBvcy83Yi80MS83YjQxZjI1Y2JiNDZkOTY4Y2YwMDAxMTA3OGI1OGMwNzIyODQxMzQ2YTNmYzNmYzY1ZGMyNGMzYTQ3MmJlM2Y1LzViNzQyODQ0MDQwNDk1ZmNiYjlkYjJhZmZlNTQ4YzVmMjhiNGEyMmFjMjdiOWEyZWEwNDdjNGRmMTlmZDE3ZTc%7EcmVzcG9uc2UtY29udGVudC1kaXNwb3NpdGlvbj0qJnJlc3BvbnNlLWNvbnRlbnQtdHlwZT0qIn1dfQ__&Signature=RLlvVNgjZD5bnFosmuhvWZyOrE-gI-j8WGCX%7E5D4wu%7EfDAuuQfMbESYSwP0Z8aK%7Eru6-XIAVtTjCXZ4xSNSQc7Y4q6ksVBnyQBSPdpC9HKiCf5ZVbegriqfUcmXBnpO8kFR8NDJDbtxZ1H-uJAsqWHJVsgz

pytorch_model.bin:  57%|#####7    | 3.30G/5.77G [00:00<?, ?B/s]

/home/reni/miniconda3/envs/hws_ws23/lib/python3.10/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  return self.fget.__get__(instance, owner)()


tokenizer_config.json:   0%|          | 0.00/905 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.89M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/1.20M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/4.80M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/582 [00:00<?, ?B/s]

  0%|          | 0/1 [00:00<?, ?it/s]

Perplexity for Uzbek text:  301.0086364746094
